File: examples/python/Depth/depth_rgb_align.py
#!/usr/bin/env python3

import numpy as np
import cv2
import depthai as dai
import time


class FPSCounter:
    def __init__(self):
        self.frameTimes = []

    def tick(self):
        now = time.time()
        self.frameTimes.append(now)
        self.frameTimes = self.frameTimes[-10:]

    def getFps(self):
        if len(self.frameTimes) <= 1:
            return 0
        return (len(self.frameTimes) - 1) / (self.frameTimes[-1] - self.frameTimes[0])


pipeline = dai.Pipeline()
device = pipeline.getDefaultDevice()

# Find a color camera automatically. Pick the first connected camera that
# advertises a COLOR sensor instead of hardcoding a socket.
colorSocket = None
for features in device.getConnectedCameraFeatures():
    if dai.CameraSensorType.COLOR in features.supportedTypes:
        colorSocket = features.socket
        break
if colorSocket is None:
    raise RuntimeError("No color camera found on this device to align the depth to.")
print(f"Aligning depth to color camera on socket: {colorSocket}")

# Unified Depth node (AUTO backend selection). The Depth node manages its own
# stereo backend and stereo cameras, resolving the algorithm at wiring time.
depthNode = pipeline.create(dai.node.Depth)

camRgb = pipeline.create(dai.node.Camera).build(colorSocket)
sync = pipeline.create(dai.node.Sync)

rgbOut = camRgb.requestOutput((640, 400), dai.ImgFrame.Type.RGB888i, enableUndistortion=True)

# Align depth to the color camera. The Depth node wires the alignment internally
# (StereoDepth's native alignTo or an ImageAlign node) depending on the platform
# and the resolved backend, so no ImageAlign node is needed.
depthNode.setAlignTo(rgbOut)
requestedOutput = depthNode.requestOutput()

rgbOut.link(sync.inputs["rgb"])
requestedOutput.depth.link(sync.inputs["depth_aligned"])

queue = sync.out.createOutputQueue()


def colorizeDepth(frameDepth):
    invalidMask = frameDepth == 0
    # Log the depth, minDepth and maxDepth
    try:
        minDepth = np.percentile(frameDepth[frameDepth != 0], 3)
        maxDepth = np.percentile(frameDepth[frameDepth != 0], 95)
        logDepth = np.zeros_like(frameDepth, dtype=np.float32)
        np.log(frameDepth, where=frameDepth != 0, out=logDepth)
        logMinDepth = np.log(minDepth)
        logMaxDepth = np.log(maxDepth)
        np.nan_to_num(logDepth, copy=False, nan=logMinDepth)
        # Clip the values to be in the 0-255 range
        logDepth = np.clip(logDepth, logMinDepth, logMaxDepth)

        # Interpolate only valid logDepth values, setting the rest based on the mask
        depthFrameColor = np.interp(logDepth, (logMinDepth, logMaxDepth), (0, 255))
        depthFrameColor = np.nan_to_num(depthFrameColor)
        depthFrameColor = depthFrameColor.astype(np.uint8)
        depthFrameColor = cv2.applyColorMap(depthFrameColor, cv2.COLORMAP_JET)
        # Set invalid depth pixels to black
        depthFrameColor[invalidMask] = 0
    except IndexError:
        # Frame is likely empty
        depthFrameColor = np.zeros((frameDepth.shape[0], frameDepth.shape[1], 3), dtype=np.uint8)
    except Exception as e:
        raise e
    return depthFrameColor


rgbWeight = 0.4
depthWeight = 0.6


def updateBlendWeights(percentRgb):
    """
    Update the rgb and depth weights used to blend depth/rgb image

    @param[in] percent_rgb The rgb weight expressed as a percentage (0..100)
    """
    global depthWeight
    global rgbWeight
    rgbWeight = float(percentRgb) / 100.0
    depthWeight = 1.0 - rgbWeight


# Connect to device and start pipeline
with pipeline:
    pipeline.build()
    print(f"Resolved algorithm: {depthNode.getResolvedAlgorithm()}")
    print(f"Resolved config:    {depthNode.getResolvedConfig()}")
    pipeline.start()

    # Configure windows; trackbar adjusts blending ratio of rgb/depth
    windowName = "rgb-depth"

    # Set the window to be resizable and the initial size
    cv2.namedWindow(windowName, cv2.WINDOW_NORMAL)
    cv2.resizeWindow(windowName, 800, 600)
    cv2.createTrackbar(
        "RGB Weight %",
        windowName,
        int(rgbWeight * 100),
        100,
        updateBlendWeights,
    )
    fpsCounter = FPSCounter()
    while True:
        messageGroup = queue.get()
        fpsCounter.tick()
        assert isinstance(messageGroup, dai.MessageGroup)
        frameRgb = messageGroup["rgb"]
        assert isinstance(frameRgb, dai.ImgFrame)
        frameDepth = messageGroup["depth_aligned"]
        assert isinstance(frameDepth, dai.ImgFrame)

        # Blend when both received
        if frameDepth is not None:
            cvFrame = frameRgb.getCvFrame()
            # Colorize the aligned depth
            alignedDepthColorized = colorizeDepth(frameDepth.getFrame())
            cv2.imshow("Depth aligned", alignedDepthColorized)

            if len(cvFrame.shape) == 2:
                cvFrame = cv2.cvtColor(cvFrame, cv2.COLOR_GRAY2BGR)
            blended = cv2.addWeighted(
                cvFrame, rgbWeight, alignedDepthColorized, depthWeight, 0
            )
            cv2.putText(
                blended,
                f"FPS: {fpsCounter.getFps():.2f}",
                (10, 30),
                cv2.FONT_HERSHEY_SIMPLEX,
                1,
                (255, 255, 255),
                2,
            )
            cv2.imshow(windowName, blended)

        key = cv2.waitKey(1)
        if key == ord("q"):
            break

File: examples/cpp/Depth/depth_rgb_align.cpp
#include <algorithm>
#include <chrono>
#include <cmath>
#include <deque>
#include <iostream>
#include <opencv2/opencv.hpp>
#include <optional>
#include <string>
#include <type_traits>
#include <variant>
#include <vector>

#include "depthai/depthai.hpp"

namespace {

const char* algorithmName(dai::node::Depth::Algorithm algorithm) {
    switch(algorithm) {
        case dai::node::Depth::Algorithm::AUTO:
            return "auto";
        case dai::node::Depth::Algorithm::STEREO:
            return "stereo";
        case dai::node::Depth::Algorithm::NEURAL:
            return "neural";
        case dai::node::Depth::Algorithm::NEURAL_ASSISTED_STEREO:
            return "neural_assisted_stereo";
        case dai::node::Depth::Algorithm::TOF:
            return "tof";
        case dai::node::Depth::Algorithm::GPU_STEREO:
            return "gpu_stereo";
    }
    return "unknown";
}

void printResolvedConfig(const dai::node::Depth::Config& config) {
    std::visit(
        [](const auto& value) {
            using T = std::decay_t<decltype(value)>;
            if constexpr(std::is_same_v<T, std::monostate>) {
                std::cout << "none";
            } else {
                std::cout << static_cast<int>(value);
            }
        },
        config);
    std::cout << std::endl;
}

class FPSCounter {
   public:
    void tick() {
        auto now = std::chrono::steady_clock::now();
        frameTimes.push_back(now);
        while(frameTimes.size() > 10) {
            frameTimes.pop_front();
        }
    }

    float getFps() const {
        if(frameTimes.size() <= 1) return 0.0f;
        auto duration = std::chrono::duration_cast<std::chrono::duration<float>>(frameTimes.back() - frameTimes.front()).count();
        return static_cast<float>(frameTimes.size() - 1) / duration;
    }

   private:
    std::deque<std::chrono::steady_clock::time_point> frameTimes;
};

float rgbWeight = 0.4f;
float depthWeight = 0.6f;

void updateBlendWeights(int percentRgb, void*) {
    rgbWeight = static_cast<float>(percentRgb) / 100.0f;
    depthWeight = 1.0f - rgbWeight;
}

}  // namespace

int main() {
    dai::Pipeline pipeline;
    auto device = pipeline.getDefaultDevice();
    if(device == nullptr) {
        throw std::runtime_error("No default device available.");
    }

    std::optional<dai::CameraBoardSocket> colorSocket;
    for(const auto& features : device->getConnectedCameraFeatures()) {
        if(std::find(features.supportedTypes.begin(), features.supportedTypes.end(), dai::CameraSensorType::COLOR) != features.supportedTypes.end()) {
            colorSocket = features.socket;
            break;
        }
    }
    if(!colorSocket.has_value()) {
        throw std::runtime_error("No color camera found on this device to align the depth to.");
    }
    std::cout << "Aligning depth to color camera on socket: " << static_cast<int>(*colorSocket) << std::endl;

    auto depthNode = pipeline.create<dai::node::Depth>();
    auto camRgb = pipeline.create<dai::node::Camera>();
    auto sync = pipeline.create<dai::node::Sync>();

    camRgb->build(*colorSocket);
    auto* rgbOut = camRgb->requestOutput(std::make_pair(640u, 400u), dai::ImgFrame::Type::RGB888i, dai::ImgResizeMode::CROP, std::nullopt, true);

    depthNode->setAlignTo(*rgbOut);

    rgbOut->link(sync->inputs["rgb"]);
    depthNode->depth().link(sync->inputs["depth_aligned"]);

    auto queue = sync->out.createOutputQueue();

    pipeline.build();

    std::cout << "Resolved algorithm: " << algorithmName(depthNode->getResolvedAlgorithm()) << std::endl;
    std::cout << "Resolved config:    ";
    printResolvedConfig(depthNode->getResolvedConfig());

    const std::string windowName = "rgb-depth";
    cv::namedWindow(windowName, cv::WINDOW_NORMAL);
    cv::resizeWindow(windowName, 800, 600);
    cv::createTrackbar("RGB Weight %", windowName, nullptr, 100, updateBlendWeights);
    cv::setTrackbarPos("RGB Weight %", windowName, static_cast<int>(rgbWeight * 100));

    FPSCounter fpsCounter;

    pipeline.start();

    while(pipeline.isRunning()) {
        auto messageGroup = queue->get<dai::MessageGroup>();
        fpsCounter.tick();

        auto frameRgb = messageGroup->get<dai::ImgFrame>("rgb");
        auto frameDepth = messageGroup->get<dai::ImgFrame>("depth_aligned");

        if(frameDepth != nullptr) {
            cv::Mat cvFrame = frameRgb->getCvFrame();
            cv::Mat alignedDepthColorized = dai::utility::colorizeDepthFrame(*frameDepth).getCvFrame();
            cv::imshow("Depth aligned", alignedDepthColorized);

            if(cvFrame.channels() == 1) {
                cv::cvtColor(cvFrame, cvFrame, cv::COLOR_GRAY2BGR);
            }

            cv::Mat blended;
            cv::addWeighted(cvFrame, rgbWeight, alignedDepthColorized, depthWeight, 0, blended);
            cv::putText(blended, "FPS: " + std::to_string(fpsCounter.getFps()), cv::Point(10, 30), cv::FONT_HERSHEY_SIMPLEX, 1, cv::Scalar(255, 255, 255), 2);
            cv::imshow(windowName, blended);
        }

        if(cv::waitKey(1) == 'q') {
            break;
        }
    }

    pipeline.stop();
    return 0;
}
